src/frontend/src/optimizer/rule/apply_scan.rs

```diff
@@ -0,0 +1,121 @@
+
+use risingwave_common::types::DataType;
+use risingwave_pb::plan_common::JoinType;
+
+use super::{BoxedRule, Rule};
+use crate::expr::{Expr, ExprImpl, ExprType, FunctionCall, InputRef};
+use crate::optimizer::plan_node::{LogicalJoin, LogicalProject, PlanTreeNodeBinary};
+use crate::optimizer::PlanRef;
+
+pub struct ApplyScan {}
+impl Rule for ApplyScan {
+    fn apply(&self, plan: PlanRef) -> Option<PlanRef> {
+        let apply = plan.as_logical_apply()?;
+        let (left, right, on, join_type) = apply.clone().decompose();
+        let apply_left_len = left.schema().len();
+        assert_eq!(join_type, JoinType::Inner);
+        // TODO: Push `LogicalApply` down `LogicalJoin`.
+        if let (None, None) = (right.as_logical_scan(), right.as_logical_join()) {
+            return None;
+        }
+
+        // Record the mapping from `CorrelatedInputRef`'s index to `InputRef`'s index.
+        // We currently can remove DAG only if ALL the `CorrelatedInputRef` are equal joined to
+        // `InputRef`.
+        // TODO: Do some transformation for IN, and try to remove DAG for it.
+        let mut column_mapping = HashMap::new();
+        on.conjunctions.iter().for_each(|expr| {
+            if let ExprImpl::FunctionCall(func_call) = expr {
+                if let Some((left, right, data_type)) = Self::check(func_call, apply_left_len) {
+                    column_mapping.insert(left, (right, data_type));
+                }
+            }
+        });
+        if column_mapping.len() == apply_left_len {
+            // Remove DAG.
+
+            // Replace `LogicalApply` with `LogicalProject` and insert the `InputRef`s which is
+            // equal to `CorrelatedInputRef` at the beginning of `LogicalProject`.
+            // See the fourth section of Unnesting Arbitrary Queries for how to do the optimization.
+            let mut exprs: Vec<ExprImpl> = (0..apply_left_len)
+                .into_iter()
+                .map(|left| {
+                    let (right, data_type) = column_mapping.get(&left).unwrap();
+                    InputRef::new(*right - apply_left_len, data_type.clone()).into()
+                })
+                .collect();
+            exprs.extend(
+                right
+                    .schema()
+                    .data_types()
+                    .into_iter()
+                    .enumerate()
+                    .map(|(index, data_type)| InputRef::new(index, data_type).into()),
+            );
+            let project = LogicalProject::create(right, exprs);
+            // TODO: add LogicalFilter here to do null-check.
+            Some(project)
+        } else {
+            // Rewrite `LogicalApply`'s left.
+
+            // Use `rewrite_agg`, `rewrite_project`, `rewrite_join` and `rewrite_scan` to remove
+            // useless columns.
+            let left = apply.left();
+            let new_left = left
+                .as_logical_agg()
+                .map(|agg| agg.rewrite_agg().unwrap())
```

Comment:
What about just call `agg.prune_col(0..agg.apply_left_len)`? It seems `rewrite_agg`, `rewrite_project`, `rewrite_join` and `rewrite_scan` are similar to prune_col and may be buggy.

Reply:
No. If one child of join output no column, we should eliminate this join and return its another child. This behavior is necessary in this case, but it's not the duty of prune_col. Besides, we can't do this JoinElimination in common case because of the unalignment with PG.

Reply:
Had an "offline" sync with @likg227 on this and there are 2 ways to proceed:

Reusing `prune_col` does produce a different result set than `rewrite_agg`, and is not the original intention of planning `AggDistinct-Project`. But it might still be correct for the overall query. We need to carefully revisit the definition of "Domain" and validate the correctness of this approach.

If we do keep the current behavior, it is important to refactor and reuse `prune_col` as much as possible. Duplicating logic can be error prone and hard to maintain - the current impl of `rewrite_scan` and `rewrite_join` is already buggy in some cases, and `prune_col` already went thru several fixes (#863 #1205 #2328 #2347 #2363 #2889)